man/mesg.1
'\" -*- coding: UTF-8 -*-
.\" Copyright (C) 1998-2001 Miquel van Smoorenburg.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
.\"
.\"{{{}}}
.\"{{{  Title
.TH MESG 1 "Feb 26, 2001" "sysvinit " "User Commands"
.\"}}}
.\"{{{  Name
.SH NAME
mesg \- control write access to your terminal
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B mesg
.RB [ y | n ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
.B mesg
controls the access to your terminal by others.  It's typically used to
allow or disallow other users to write to your terminal (see \fBwrite\fP(1)).
.\"}}}
.\"{{{  Options
.SH OPTIONS
.IP \fBy\fP
Allow write access to your terminal.
.IP \fBn\fP
Disallow write access to your terminal.
.PP
If no option is given, \fBmesg\fP prints out the current access state of your
terminal.
.\"}}}
.\"{{{  Notes
.SH NOTES
\fBmesg\fP assumes that its standard input is connected to your
terminal. That also means that if you are logged in multiple times,
you can get/set the mesg status of other sessions by using redirection.
For example "mesg n < /dev/pts/46".
.SH AUTHOR
.MT miquels@\:cistron\:.nl
Miquel van Smoorenburg
.ME
.\"}}}
.\"{{{  See also
.SH "SEE ALSO"
.BR talk (1),
.BR write (1),
.BR wall (1)
.\"}}}


doc/Changelog
sysvinit (3.03) unreleased; urgency=low
    * Fixed formatting in init.8 man page.
      Patch provided by Mark Hindley.
    * Fixed compile error on GNU Hurd due to missing MAX_PATH definition.
      Error reported by Mark Hindly.

sysvinit (3.02) released; urgency=low
    * Added q and Q flags to synopsis in shutdown manual page.
    * Applied fixes for markup and spacing in manual pages.
      Patch provided by Mario Blattermann.
    * Added translation framework (po4a) from Mario Blttermann.
    * Added Makefile for man/ directory. Will handle translations
      and substitutions.
    * Applied new translations for multiple languages from Mario Blattermann.
    * Added ability to use "@" symbol in command named in the inittab file. This
      treats commands as literal and does not launch a shell to interpret them.
    * Updated inittab manual page to include overview of symbols which trigger
      a shell interpretor and how to disable them using the @ symbol.
    * Introduced change which adds error checking in bootlogd when performing
      chdir(). - Provided by Alexander Vickberg 
    * Add check for console using TIOCGDEV on Linux systems in bootlogd to
      make finding console more robust. - Provided by Alexander Vickberg
 

sysvinit (3.01) released; urgency=low
    * Default to showing processes in the uninterruptable state (D).
    The -z flag no longer affects whether processes in D state are shown.
    The -z flag does still toggle whether zombie (Z) processes are shown.

    * Removed unnecessary check which is always true from init tab parsing.

sysvinit (3.00) released; urgency=low

    * Applied patch from Matthias Schiffer which allows bootlogd to read from
      a wider range of consoles. The console name is already passed in from the
      kernel command line using "console=". We no longer filter out names as strictly
      but do now check to confirm the "console=" device points to a valid TTY.

sysvinit (2.99) released; urgency=low

    * Fixed typos and missing underlines in shutdown manual page.
      Corrections provided by Helge Kreutzmann.

sysvinit (2.98) released; urgency=low

    * Fixed time parsing in shutdown when there is a + in front of a 0 time offset.
      Commands with a positive time offset (+1) would work but +0 fails.
      This has been corrected by Arkadiusz Miskiewicz.

sysvinit (2.97) released; urgency=low

    * Check $(ROOT) filesystem for libcrypt instead of a hardcoded path to /usr.
      Added logsave and readbootlog to list of files git ignores.
      - Patches provided by Petr Ovtchenkov.

    * Code clean-up and making sure we avoid freeing unused memory.
      Patch provided by David Hoyer.

    * Added shell script from Trek which converts systemd unit files
      into init.d style scripts.

    * Added patch from Didier Gaudin which allows init to load configuration
      data from files stored in /etc/inittab.d/

    * Added patch from William Shipley which allows shutdown time to be specified
      in the format +hh:mm. This is in addition to the existing formats such as
      hh:mm, +m, and "now".

    * Fixed typos in manual pages. Submitted by Helge Kreutzmann.


sysvinit (2.96) released; urgency=low

  [ Jesse Smith ]

    * Added -z command line parameter to pidof which tells pidof to
      try to find processes in uninterruptible (D) or zombie (Z) states.
      This can cause pidof to hang, but produces a more complete process
      list.
      Closes Savannah bug #56534

    * Reformatted init code to make if/while logic more clear.

    * Cleaned up some output from readbootlog.

    * Added -e flag to bootlogd. When -e is used, data saved
      to the boot log file does not have escape characters
      removed. This means colour and cursor movement codes
      stay in the log file. The may then look nicer when
      read with "less -R', but may appear cluttered or
      out of alignment when viewed with other, plain-text tools.
      When -e is not used, escape characters are removed/filtered.
      Closes Debian bug #672361.

    * Make sure src/Makefile cleans up all executable files
      when parent Makefile calls "make clean".


sysvinit (2.95) released; urgency=low

  [ Jesse Smith ]

    * Fixed various typos and control codes in manual pages.
      Patch provided by Bjarni Ingi Gislason.
    * Dropped "-f" format flag for pidof command as it
      could be used to print information from memory or crash
      pidof. Replaced flag with "-d" which allows for a custom
      separator to be used between listed PIDs.
      Patch supplied by KatolaZ.
    * Updated manual page to describe -h and its modifiers (-H and -P)
      in more detail. Should close Debian bug #374039.
    * Use millisecond delays in init so that shutdown
      can happen without a near-full-second delay after
      all processes have terminated. Replaced do_sleep()
      with do_msleep(), provided by Serge Belyshev.
    * Replaced hardwired sleep constants in init.c with
      defined constants for easy updating in the future.
    * Accepted patch from Luc Gommans which explains why zombie and
      deep sleep processes (Z and D) may be hidden from pidof.
    * Removed link instruction against sepol library from src/Makefile.
      This has no change to the resulting binary, just makes the
      Makefile cleaner.

sysvinit (2.94) released; urgency=low

  [ Jesse Smith ]

    * When the halt command is called with the -p flag (or as poweroff)
      the command now passes the "-h -P" flags to shutdown. This
      in turn sets the INIT_HALT environment variable to POWEROFF.
      Assuming this value is checked by initscripts during the
      shutting down procedure, it should cause the system to
      be powered off.
      If halt is called without -p then the value of INIT_HALT
      is not set and the default action (often set in /etc/default/halt)
      is taken.
    * Removed unnecessary malloc.h includes. Memory allocation

src/mesg.c
/*
 * mesg.c	The "mesg" utility. Gives / restrict access to
 *		your terminal by others.
 *
 * Usage:	mesg [y|n].
 *		Without arguments prints out the current settings.
 *
 *		This file is part of the sysvinit suite,
 *		Copyright (C) 1991-2001 Miquel van Smoorenburg.
 *
 *		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 2 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License
 *		along with this program; if not, write to the Free Software
 *		Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <grp.h>

char *Version = "@(#) mesg 2.81 31-Jul-2001 miquels@cistron.nl";

#define TTYGRP		"tty"

/*
 *	See if the system has a special 'tty' group.
 *	If it does, and the tty device is in that group,
 *	we set the modes to -rw--w--- instead if -rw--w--w.
 */
int hasttygrp(void)
{
	struct group *grp;

	if ((grp = getgrnam(TTYGRP)) != NULL)
		return 1;
	return 0;
}


/*
 *	See if the tty devices group is indeed 'tty'
 */
int tty_in_ttygrp(struct stat *st)
{
	struct group *gr;

	if ((gr = getgrgid(st->st_gid)) == NULL)
		return 0;
	if (strcmp(gr->gr_name, TTYGRP) != 0)
		return 0;

	return 1;
}

int main(int argc, char **argv)
{
	struct stat	st;
	unsigned int	ttymode, st_mode_old;
	int		ht;
	int		it;
	int		e;

	if (!isatty(0)) {
		/* Or should we look in /var/run/utmp? */
		fprintf(stderr, "stdin: is not a tty\n");
		return(1);
	}

	if (fstat(0, &st) < 0) {
		perror("fstat");
		return(1);
	}

	ht = hasttygrp();
	it = tty_in_ttygrp(&st);

	if (argc < 2) {
		ttymode = (ht && it) ? 020 : 002;
		printf("is %s\n", (st.st_mode & ttymode) ? "y" : "n");
		return 0;
	}
	if (argc > 2 || (argv[1][0] != 'y' && argv[1][0] != 'n')) {
		fprintf(stderr, "Usage: mesg [y|n]\n");
		return 1;
	}

	/*
	 *	Security check: allow mesg n when group is
	 *	weird, but don't allow mesg y.
	 */
	ttymode = ht ? 020 : 022;
	if (ht && !it && argv[1][0] == 'y') {
		fprintf(stderr, "mesg: error: tty device is not owned "
			"by group `%s'\n", TTYGRP);
		exit(1);
	}

	st_mode_old = st.st_mode;
	if (argv[1][0] == 'y')
		st.st_mode |= ttymode;
	else
		st.st_mode &= ~(ttymode);
	if (st_mode_old != st.st_mode && fchmod(0, st.st_mode) != 0) {
		e = errno;
		fprintf(stderr, "mesg: %s: %s\n",
			ttyname(0), strerror(e));
		exit(1);
	}

	return 0;
}
